presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java

```diff
@@ -291,6 +415,98 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(S
+                .extraParameters(MathFunctions::decimalRoundNExtraParameters)
+                .build();
+    }
+
+    private static SqlScalarFunction decimalRoundFunction()
+    {
+        Signature signature = Signature.builder()
+                .kind(SCALAR)
+                .name("round")
+                .literalParameters("num_precision", "num_scale")
+                .argumentTypes("decimal(num_precision, num_scale)")
+                .returnType("decimal(num_precision, num_scale)")
+                .build();
+        return SqlScalarFunction.builder(MathFunctions.class)
+                .signature(signature)
+                .description("remainder of given quotient")
+                .methods("roundShortDecimal", "roundLongDecimal")
+                .extraParameters(MathFunctions::decimalRoundExtraParameters)
+                .build();
+    }
+
+    private static List<Object> decimalRoundExtraParameters(SpecializeContext specializeContext)
+    {
+        return ImmutableList.of(0, decimalRoundNExtraParameters(specializeContext).get(0), decimalRoundNExtraParameters(specializeContext).get(1));
+    }
+
+    private static List<Object> decimalRoundNExtraParameters(SpecializeContext context)
+    {
+        return ImmutableList.of(context.getLiteral("num_precision"), context.getLiteral("num_scale"));
+    }
+
+    public static long roundShortDecimal(long num, long roundScale, long inputPrecision, long inputScale)
+
+    {
+        if (num == 0 || inputPrecision - inputScale + roundScale <= 0) {
+            return 0;
+        }
+        if (roundScale >= inputScale) {
+            return num;
+        }
+        if (num < 0) {
+            return -roundShortDecimal(-num, roundScale, inputPrecision, inputScale);
+        }
+
+        long rescaleFactor = tenToNth((int) (inputScale - roundScale)).longValueExact();
+        long remainder = num % rescaleFactor;
+        return (num / rescaleFactor + ((firstDigit(remainder) > 4) ? 1 : 0)) * rescaleFactor;
+    }
+
+    private static long firstDigit(long num)
+    {
+        while (num > 9) {
+            num /= 10;
+        }
+        return Math.abs(num);
+    }
+
+    public static Slice roundLongDecimal(Slice num, long roundScale, long inputPrecision, long inputScale)
+    {
+        BigInteger unscaledVal = unscaledValueToBigInteger(num);
+        if (unscaledVal.signum() == 0) {
+            return LongDecimalType.unscaledValueToSlice(0);
+        }
+        if (roundScale >= inputScale) {
+            return num;
+        }
+        BigInteger rescaleFactor = tenToNth((int) (inputScale - roundScale));
+        if (unscaledVal.signum() < 0) {
+            return LongDecimalType.unscaledValueToSlice(roundLongDecimal(unscaledVal.negate(), rescaleFactor).negate());
+        }
+        return LongDecimalType.unscaledValueToSlice(roundLongDecimal(unscaledVal, rescaleFactor));
+    }
+
+    public static BigInteger roundLongDecimal(BigInteger num, BigInteger rescaleFactor)
```

Comment:
change it so it uses exactly same arithmetic operations as `long` based version so it is easier to follow.